models/migrations/v999.go

```diff
@@ -0,0 +1,28 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"xorm.io/xorm"
+)
+
+func creatUserBadgesTable(x *xorm.Engine) error {
+	type Badge struct {
+		ID          int64 `xorm:"pk autoincr"`
+		Description string
+		ImageURL    string
+	}
+
+	type userBadge struct {
+		ID      int64 `xorm:"pk autoincr"`
+		BadgeID int64
+		UserID  int64
```

Comment:
UserID should have index

Reply:
Every user should have only one record?

Reply:
No but mostly records will be queried by UserID to get user badges

Reply:
OK before we start adding indices we should check what queries there are.

## Current Queries

Looking at the file models/user/badge.go there appears to be only one query at present: 

```sql
SELECT badge.* INNER JOIN user_badge ON badge.id = user_badge.badge_id WHERE user_badge.user_id = ?
```

This would suggest that an index on `user_badge.user_id` is required and that a standalone `badge_id` index isn't of any use. 

In fact the existence of the `badge_id` index would likely result in it being used in preference to that of one on `user_id` even though a single `user_id` index would be in almost all likely scenarios.

## Composite Indices?

Now is it worth creating a composite index containing both `user_badge.user_id` and  `user_badge.badge_id`?

Likely yes - but the index order will need to be considered.

The reasoning is as follows:

* Although I'm not sure which db can use the `badge_id` information straight from the composite index some will. In those cases that would make the inner join essentially a pure index lookup without requiring a second lookup in to the table.
* It is likely that future queries will also want to restrict by `badge_id` at the very least during deletion. Now for safety, most (if not all) queries should restrict by `user_id`. (And it is likely any that are not doing probably won't need to be fast.)

### So what is the best order?

As I suspect that all queries should be restricted by `user_id` and that some db systems will benefit from having the `badge_id` in the index then I think:

`INDEX(user_id, badge_id)`

is going to be the best index.

## Is it worth adding a single `badge_id` index?

Adding an index to `badge_id` by itself will very likely cause the above `INNER JOIN` query to choose the`badge_id` index instead of `user_id` even if a composite index is present. So DON'T do it.

## ASIDE: Should we actually query on `user_badge` instead `badge`?

Proper testing is required here with the above index present, but it may be worth reconstituting this query:

```sql
SELECT badge.* FROM badge INNER JOIN user_badge ON badge.id = user_badge.badge_id WHERE user_badge.user_id = ?
```

into:

```sql
SELECT badge.* FROM user_badge INNER JOIN badge ON badge.id = user_badge.badge_id WHERE user_badge.user_id = ?
```

Why would we do this? Because of the potential for the composite index to prevent lookup in the `user_badge` table. We should check if one or the other of these two queries encourages that behaviour. The SQL optimisers should be doing this for us but it is worth checking.